### PR TITLE
Enhancement: Field data export

### PR DESCRIPTION
**Pull request by MuellerSeb**
This PR adresses #15. In addition I made some corrections in the code:
- reformatted with black
- removed call to non-existing `writeBlockSize64Bit`
- removed `largeFile` attribute form `VtkFile`
- rewrote doc string in numpydoc style
- converted None comparisons from `val != None` to `val is not None`

### CODE REVIEW

**Comment by renefritze** [comment]
Thank you for the contribution @MuellerSeb !

I'm having a little trouble discovering the functional changes in between the formatting ones. Can you help me with that?

I've also update the CI setup in #18 just now, if you rebase on master now, travis can work again.

File: examples/unstructured.py [hunk 1]
```diff
@@ -48,11 +48,14 @@
 x[4], y[4], z[4] = 1.0, 1.0, 0.0
 x[5], y[5], z[5] = 2.0, 1.0, 0.0
 
+point_data = {"test_pd": np.array([1, 2, 3, 4, 5, 6])}
+cell_data = {"test_cd": np.array([1, 2, 3])}
+field_data = {"test_fd": np.array([1.0, 2.0])}
```
Comment (MuellerSeb) [inline examples/unstructured.py:53]: Here I added field data to the grid.

File: examples/unstructured.py [hunk 2]
```diff
@@ -67,4 +70,15 @@
...
+    "unstructured",
+    x,
+    y,
+    z,
+    connectivity=conn,
+    offsets=offset,
+    cell_types=ctype,
+    cellData=cell_data,
+    pointData=point_data,
+    fieldData=field_data,
```
Comment (MuellerSeb) [inline examples/unstructured.py:83]: This field data can now simply be added to the call of the export functions.

File: pyevtk/hl.py [hunk 3]
```diff
@@ -22,23 +22,30 @@
...
+    VtkPolyLine,
+    VtkPixel,
+)
+
 
 # =================================
 #       Helper functions
 # =================================
-def _addDataToFile(vtkFile, cellData, pointData):
+def _addDataToFile(vtkFile, cellData, pointData, fieldData=None):
```
Comment (MuellerSeb) [inline pyevtk/hl.py:44]: Here you can now pass field data to create a corresponding section.

File: pyevtk/hl.py [hunk 4]
```diff
@@ -47,132 +54,197 @@ def _addDataToFile(vtkFile, cellData, pointData):
...
+        vtkFile.openData("Cell", scalars=keys[0])
         for key in keys:
             data = cellData[key]
             vtkFile.addData(key, data)
         vtkFile.closeData("Cell")
 
-def _appendDataToFile(vtkFile, cellData, pointData):
+    # Field data
+    # https://www.visitusers.org/index.php?title=Time_and_Cycle_in_VTK_files#XML_VTK_files
+    if fieldData:
```
Comment (MuellerSeb) [inline pyevtk/hl.py:65]: Field data has no attributes, so just a <FieldData> section is created.

File: pyevtk/hl.py [hunk 5]
```diff
@@ -47,132 +54,197 @@ def _addDataToFile(vtkFile, cellData, pointData):
...
+        keys = list(fieldData.keys())
+        vtkFile.openData("Field")  # no attributes in FieldData
+        for key in keys:
+            data = fieldData[key]
+            vtkFile.addData(key, data)
+        vtkFile.closeData("Field")
+
+
+def _appendDataToFile(vtkFile, cellData, pointData, fieldData=None):
```
Comment (MuellerSeb) [inline pyevtk/hl.py:74]: Here we can now create DataSets for field data.

File: pyevtk/hl.py [hunk 6]
```diff
@@ -47,132 +54,197 @@ def _addDataToFile(vtkFile, cellData, pointData):
...
-            Full path to saved file.
-
-        NOTE: At least, cellData or pointData must be present to infer the dimensions of the image.
+def imageToVTK(
+    path,
+    origin=(0.0, 0.0, 0.0),
+    spacing=(1.0, 1.0, 1.0),
+    cellData=None,
+    pointData=None,
+    fieldData=None,
```
Comment (MuellerSeb) [inline pyevtk/hl.py:104]: Added `fieldData` keyword to `imageToVTK`

File: pyevtk/hl.py [hunk 7]
```diff
@@ -47,132 +54,197 @@ def _addDataToFile(vtkFile, cellData, pointData):
...
     w.closeGrid()
-    _appendDataToFile(w, cellData, pointData)
+    _appendDataToFile(w, cellData, pointData, fieldData)
     w.save()
     return w.getFileName()
 
+
 # ==============================================================================
-def gridToVTK(path, x, y, z, cellData = None, pointData = None):
+def gridToVTK(path, x, y, z, cellData=None, pointData=None, fieldData=None):
```
Comment (MuellerSeb) [inline pyevtk/hl.py:181]: Added `fieldData` keyword to `gridToVTK`

File: pyevtk/hl.py [hunk 8]
```diff
@@ -182,282 +254,422 @@ def gridToVTK(path, x, y, z, cellData = None, pointData = None):
...
     # Write data
-    _appendDataToFile(w, cellData, pointData)
+    _appendDataToFile(w, cellData, pointData, fieldData)
     w.save()
     return w.getFileName()
 
 
 # ==============================================================================
-def pointsToVTK(path, x, y, z, data = None):
+def pointsToVTK(path, x, y, z, data=None, fieldData=None):
```
Comment (MuellerSeb) [inline pyevtk/hl.py:275]: Added `fieldData` keyword to `pointsToVTK`

File: pyevtk/hl.py [hunk 9]
```diff
@@ -182,282 +254,422 @@ def gridToVTK(path, x, y, z, cellData = None, pointData = None):
...
-    _appendDataToFile(w, cellData = None, pointData = data)
+    _appendDataToFile(w, cellData=None, pointData=data, fieldData=fieldData)
 
     w.save()
     return w.getFileName()
 
+
 # ==============================================================================
-def linesToVTK(path, x, y, z, cellData = None, pointData = None):
+def linesToVTK(path, x, y, z, cellData=None, pointData=None, fieldData=None):
```
Comment (MuellerSeb) [inline pyevtk/hl.py:343]: Added `fieldData` keyword to `linesToVTK`

File: pyevtk/hl.py [hunk 10]
```diff
@@ -182,282 +254,422 @@ def gridToVTK(path, x, y, z, cellData = None, pointData = None):
...
+    )
 
     w.save()
     return w.getFileName()
 
+
 # ==============================================================================
-def polyLinesToVTK(path, x, y, z, pointsPerLine, cellData = None, pointData = None):
+def polyLinesToVTK(
+    path, x, y, z, pointsPerLine, cellData=None, pointData=None, fieldData=None
```
Comment (MuellerSeb) [inline pyevtk/hl.py:431]: Added `fieldData` keyword to `polyLinesToVTK`

File: pyevtk/hl.py [hunk 11]
```diff
@@ -182,282 +254,422 @@ def gridToVTK(path, x, y, z, cellData = None, pointData = None):
...
+    path,
+    x,
+    y,
+    z,
+    connectivity,
+    offsets,
+    cell_types,
+    cellData=None,
+    pointData=None,
+    fieldData=None,
```
Comment (MuellerSeb) [inline pyevtk/hl.py:526]: Added `fieldData` keyword to `unstructuredGridToVTK`

File: pyevtk/hl.py [hunk 12]
```diff
@@ -182,282 +254,422 @@ def gridToVTK(path, x, y, z, cellData = None, pointData = None):
...
+    x0,
+    y0,
+    z0,
+    z1,
+    radius,
+    nlayers,
+    npilars=16,
+    cellData=None,
+    pointData=None,
+    fieldData=None,
```
Comment (MuellerSeb) [inline pyevtk/hl.py:621]: Added `fieldData` keyword to `cylinderToVTK`

File: pyevtk/vtk.py [hunk 13]
```diff
@@ -116,271 +151,387 @@ def __str__(self):
...
+    filepath : str
+        filename without extension.
+    ftype : str
+        file type, e.g. VtkImageData, etc.
+    largeFile : bool, optional
+        If size of the stored data cannot be represented by a UInt32.
+        The default is False.
+    """
+
+    def __init__(self, filepath, ftype):
```
Comment (MuellerSeb) [inline pyevtk/vtk.py:259]: removed the unused `largeFile` keyword.

File: pyevtk/vtk.py [hunk 14]
```diff
@@ -116,271 +151,387 @@ def __str__(self):
...
+            default data array name for tensors data.
+        tcoords : str, optional
+            default data array name for tcoords data.
+
+        Returns
+        -------
+        VtkFile
+            This VtkFile to allow chained calls.
         """
         self.xml.openElement(nodeType + "Data")
```
Comment (MuellerSeb) [inline pyevtk/vtk.py:388]: This creates the FieldData element, when `nodeType="Field"`

File: pyevtk/vtk.py [hunk 15]
```diff
@@ -392,96 +543,124 @@ def addData(self, name, data):
...
+        ncomp : int
+            number of components, 1 (=scalar) or 3 (=vector)..
         """
         self.openAppendedData()
         dsize = np_to_vtk[dtype].size
         block_size = dsize * ncomp * nelem
-        if self.largeFile == False:
-            writeBlockSize(self.xml.stream, block_size)
-        else:
-            writeBlockSize64Bit(self.xml.stream, block_size)
```
Comment (MuellerSeb) [inline pyevtk/vtk.py:409]: This routine did not exists.

**Comment by MuellerSeb** [comment]
> Thank you for the contribution @MuellerSeb !
> 
> I'm having a little trouble discovering the functional changes in between the formatting ones. Can you help me with that?
> 
> I've also update the CI setup in #18 just now, if you rebase on master now, travis can work again.

I did a review, pointing out the essential changes.

I also rebased and travis succeeded.

**Comment by renefritze** [comment]
Great! Thanks. Those changes look good to me.

While I do not mind the style changes personally, lets see if @somada141 or @xylar have strong opinions.

**Comment by MuellerSeb** [comment]
Cool. Is @somada141 still active on GitHub?

**Comment by renefritze** [comment]
> I like the updated formatting. It might be hard to merge changes from Paulo Herrera's version but I don't think we expect much from there.

Indeed. There hasn't been anything in over a year. We can always just run black on the remote too before a hypothetical merge though.

> Cool. Is @somada141 still active on GitHub?

Let's wait till monday with the merge in either way.

**Comment by renefritze** [comment]
Thank you again @MuellerSeb, I'll try to prepare a new release today.

**Comment by MuellerSeb** [comment]
@renefritze : maybe we could address #19 first, before releasing a new version.

**Pull request merged**